python/paddle/fluid/dygraph/dygraph_to_static/cache_program.py

```diff
@@ -0,0 +1,292 @@
+        # it means layers(or Ops) are added into default_main_program switched by outer
+        # `with` statement.
+        self.program = framework.default_main_program()
+
+    def _prepare(self, args, kwargs):
+        """
+        Prepares batch_data and adds `fluid.data` layers into program.
+        """
+        if not self.inputs:
+            self._add_feed_layers(args, kwargs)
+        elif self.traced_funcs[-1] == self.forward_func:
+            self._update_batch_data(args, kwargs)
+
+    def add_layers(self, dyfunc):
+        """
+        Transforms dygraph function into static function, and adds layers into
+        main_program.
+        """
+        self._check_cache_valid()
+        static_func = self.func_cache(dyfunc)
+        # self._forward_func is entry function of Net or Model.
+        # It can be called for multiple times, but layers from these functions
+        # call stack will be added into self._program only once.
+        # After that, cached program will be always returned by default.
+        if static_func == self.forward_func:
+            self.is_repeated = True
+
+        if self.forward_func is None:
+            self.forward_func = static_func
+
+        self.traced_funcs.append(static_func)
+
+        return static_func
+
+    def _get_or_build_program(self, func, args, kwargs):
+        """
+        Returns program of the input function. If called at first time,
+        builds a new program and caches it.
+        """
+        with framework.program_guard(self.program):
+            if func == self.forward_func and self.is_build_process:
+                # Replaces input data with `layers.data`
+                feed_name_to_idx = self._feed_name_to_idx(self.forward_func)
+                args = list(args)
+                for feed_layer in self.inputs:
+                    idx = feed_name_to_idx[feed_layer.name]
+                    args[idx] = feed_layer
+                self.is_build_process = False
+                fetch_list = func(*args, **kwargs)
+                self._add_optimizer(fetch_list)
+            else:
+                fetch_list = func(*args, **kwargs)
+        self.outputs = fetch_list
+        return fetch_list
+
+    def _run(self, args, kwargs):
+        """
+        Executes the main_program and returns Tensors of fetch_list
+        """
+        assert self.inputs, "inputs is not initialized."
+        assert self.batch_data, "batch_data is empty."
+
+        input_names = [in_var.name for in_var in self.inputs]
+        exe = executor.Executor(core.CPUPlace())
+        if self.need_startup:
+            self.need_startup = False
+            exe.run(framework.default_startup_program())
+        feed_dict = dict(zip(input_names, self.batch_data))
+        res = exe.run(self.program, feed=feed_dict, fetch_list=self.outputs)
+        return res
+
+    def set_optimizer(self, optimizer, force_update=True):
```

Comment:
As we discussed in our group chat with Yamei, this part `set_optimizer` and `add_optimizer` is not natural. For Paddle user, they may just know program, and when calling your program cache to set_optimizer, 1. they may feel strange. 2. Here you still didn't set optimizer to real program, it's more like a "pre-set".

Let's wait you to move run logic out or cache and we will see you how to handle this case.

Reply:
That's right. Here we need a better way to expose the interface that allows users to set optimizer easily. I think maybe we can bind the `set_optimizier` to params `self` of decorator. I will continue to explore new approaches.